Python support

mattinbits opened the pull request:
Implementation of python support for issue #196

Supports SparkContext, SQLContext and HiveContext.

Adds support to JobSqlDAO for binaries of different types. Note that JobFileDAO will not work
with Python as it has not had support for different binary types added to it in this PR.

Python support does not include support for NamedObjects in this PR.

velvia commented on doc/python.md:
```diff
@@ -0,0 +1,237 @@
+# Python Support
+
+Spark Job Server supports Python jobs through a Python specific context factory 
```
Docs in the PR!   👍

velvia commented on doc/python.md:
```diff
@@ -0,0 +1,237 @@
+
+In development, you can run a Python ready job server locally with the following `reStart` task:
+
+    job-server-extras/reStart ../job-server-python/config/python-example.conf.template
```
Hmmm... not a biggie but I'd rather consolidate the Python configs into the same config/ dir as everything else

velvia commented on job-server-api/src/spark.jobserver/JobCache.scala:
```diff
@@ -2,16 +2,30 @@ package spark.jobserver
+    * @param classPath the fully qualified name of the class/object to load
+    * @return The case class containing the location of the binary file for the specified job.
+    */
+  def getPythonJob(appName: String, uploadTime: DateTime, classPath: String): PythonJobInfo
```
I wonder if this could be consolidated to a single method returning BinaryJobInfo.... will see when I look at other classes

velvia commented on job-server-extras/src/spark.jobserver/python/PythonContextFactory.scala:
```diff
@@ -0,0 +1,218 @@
+  protected def pythonPaths(config: Config): Seq[String] = {
+    val envPaths = sys.env.get("PYTHONPATH").map(_.split(":").toSeq).getOrElse(Seq())
+    val configPaths =
+      if (config.hasPath("python.paths")) {
```
We should really use this library: net.ceedubs.ficus.  It would allow us to do `config.as[Option[Seq[String]]("python.paths")`

mattinbits replied:
Alternatively, could add `python.paths = []` and  `python.executable = "python"` to application.conf so they're always present, I think this is actually the recommended way for TypeSafe config, to have a default for every possible option so its not necessary to handle the possibility of an absent property?

noorul commented on job-server-python/src/python/sparkjobserver/subprocess.py:
```diff
@@ -0,0 +1,69 @@
+from pyspark.sql import SQLContext, HiveContext
+from sparkjobserver.api import ValidationProblem
+
+def exitWithFailure(message, exitCode = 1):
```
Shall we have consistency and not use camel cases for method names and variables.

noorul commented:
I think it is very important to make tests pass.

Also I think we should run python tests using travis.

mattinbits commented:
@noorul, I completely agree re: Tests passing. Is this something you can help with? I think it needs access to the travis system which I don't have.

noorul commented:
@mattinbits I am not sure what help you need. As far as travis is concerned, I think you need to make changes to .travis.yaml and play around. Or are you looking for any specific help which needs you to log into travis as an owner of this project?

noorul commented:
@mattinbits I think you need to re-base and push for travis to work. There is a conflict.

mattinbits commented:
That's what I'm doing right now :)

mattinbits commented:
@noorul @velvia Are you waiting on any further amendments from me for this to be mergable? Sorry to be a PITA but I'm going to run out of time to keep up with the rebasing as other things that conflict get merged.

noorul commented:
@mattinbits Sorry for the delay. I have not found much time to do thorough review. I will get to this as soon as I get some time. Maybe someone else would do this early.

velvia commented on .travis.yml:
```diff
@@ -1,14 +1,22 @@
    - oraclejdk7
+before_install:
+  - sudo apt-get -qq update
+  - sudo apt-get -y install python3 python3-pip
```
Just curious... are only Py3 jobs supported?

mattinbits replied:
No, both Python2 and 3 are supported. But the image already contains python2 and pip2. Note that `ci/install-python-dependencies.sh` installs pyhocon for both. There are tests for both in `SubprocessSpec.scala`. 

The reason for changing the `dist` in the Travis file is also related to this, the older default Travis image is Ubuntu 12.04 for which installing Python3 and pip3 is awkward: http://askubuntu.com/questions/412178/how-to-install-pip-for-python-3-in-ubuntu-12-04-lts

velvia commented on doc/python.md:
```diff
@@ -0,0 +1,237 @@
+
+The interface to conform to is shown by the `SparkJob` python class:
+
+    class SparkJob:
```
You could specify Python code highlighting in Markdown with triple backtickts and then python....

velvia commented on job-server-extras/src/spark.jobserver/python/PythonContextFactory.scala:
```diff
@@ -0,0 +1,201 @@
+    val sc = new SparkContext(sparkConf)
+
+    val jsc = new JavaSparkContext(sc) with PythonContextLike with DefaultContextLikeImplementations {
+
```
(minor) could get rid of extra blank lines

velvia commented on job-server-extras/src/spark.jobserver/python/PythonJob.scala:
```diff
@@ -0,0 +1,148 @@
+        case (Success(_), (k, f@Failure(ex: Throwable))) => Failure(ex)
+        case (Success(m), (k, Success(v))) => Success(m + ((k, v)))
+      }
+    case l: java.util.ArrayList[_] =>
```
Is it always an `ArrayList`?  Would just abstracting to `java.util.List` be better?

velvia commented on job-server-extras/src/spark.jobserver/python/PythonJob.scala:
```diff
@@ -0,0 +1,148 @@
+    case m: java.util.HashMap[_,_] =>
+      val scalaMap = m.asScala.toMap
+      val recursedMap = scalaMap.mapValues(convertRawResult)
+      recursedMap.foldLeft(Success(Map.empty[Any, Any]): Try[Map[Any, Any]]) {
```
Neat, recursive functional validation of conversion.   There must be some library that does this already, but I can't think of anything off the top of my head.   

This might be expensive if the result set is fairly large.  I'd be OK with just a dumb try/catch solution too.

mattinbits replied:
The trouble is, the recursion is used to convert from java collections to scala collections, because spray-json can, by default, serialize scala collections, and the basic requirement for output is that spray can serialize it. A more efficient alternative could be to just implement a spray-json protocol for `java.util.List` and `java.util.HashMap`, and leave the return type from Python unchecked and unaltered until Spray comes to serialize it. Then we traverse the data structure only once during serialization. I suppose this would be closer to the behaviour of Scala/Java context factories, where the output is similarly assumed to be serializable.

Let me know if this is preferable and I will update the PR accordingly.

velvia commented on job-server-extras/src/spark.jobserver/python/PythonJob.scala:
```diff
@@ -0,0 +1,148 @@
+  /**
+    *
+    * To support a useful validate method here for Python jobs we would have call two python processes,
+    * one for validate and one for runJob. However this is inefficient and it would mean having to convert
```
You mean not calling two separate processes, but doing an RPC twice for the same process right?

velvia commented on job-server-python/src/python/sparkjobserver/subprocess.py:
```diff
@@ -0,0 +1,66 @@
+    c = getattr(module, class_name)
+    return c
+
+if __name__ == "__main__":
```
Should we have something like pydoc or whatever is the standard for Python app documentation?

velvia commented on job-server-python/test/spark/jobserver/python/SubprocessSpec.scala:
```diff
@@ -0,0 +1,356 @@
+    set("spark.sql.shuffle.partitions", "5")
+  lazy val sc = new SparkContext(conf)
+  lazy val jsc = new JavaSparkContext(sc)
+    with IdentifiedContext{ def contextType = classOf[JavaSparkContext].getCanonicalName}
```
space before { and } ?

velvia commented on job-server/src/spark.jobserver/JobCache.scala:
```diff
@@ -22,28 +26,51 @@ class JobCacheImpl(maxEntries: Int,
-      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
-      JobJarInfo(constructor, classPath, jarFilePath)
-    })
+    cache.get((appName, uploadTime, classPath)) match {
```
Wait a minute, why did you replace the original logic with a match?    In the original logic, the second argument to the cache get method made sure that the get is atomic:  if the item is not there, then the logic is executed and anybody else calling the get method has to wait.  Your logic OTOH is not atomic.  Please can we go back to the original logic of passing the function as the second argument?

hntd187 replied:
Ideally, we should rewrite the cache entirely to be future based and just pass futures around on this kind of stuff.

velvia replied:
The cache is already futures based!   Read up on the spray cache.   That's
why I recommend sticking with the existing method of passing in a function
as the second argument of get(...)....  which creates the future to
populate the cache if the key cannot be found, and ensures other threads
gets that same future to avoid the stampeding herds problem.

On Thu, Aug 11, 2016 at 4:14 PM, Stephen Carman <notifications@github.com>
wrote:

> In job-server/src/spark.jobserver/JobCache.scala
> <https://github.com/spark-jobserver/spark-jobserver/pull/553#discussion_r74520877>
> :
>
> >     * @param classPath the fully qualified name of the class/object to load
> >     */
> >    def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
> > -    cache.get((appName, uploadTime, classPath), {
> > -      import akka.pattern.ask
> > -      import scala.concurrent.Await
> > -
> > -      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
> > -      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
> > -      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
> > -      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
> > -      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
> > -      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
> > -      JobJarInfo(constructor, classPath, jarFilePath)
> > -    })
> > +    cache.get((appName, uploadTime, classPath)) match {
>
> Ideally, we should rewrite the cache entirely to be future based and just
> pass futures around on this kind of stuff.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/553/files/b4ea3b476442b188582c22f86fe5e18e504de3b4#r74520877>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABA323Y64ogKrm13dtU8bb-TTN8b4I7gks5qe6zogaJpZM4JZq4L>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

hntd187 replied:
But we aren't using the Spray caching we are using our own impl of the LRUCache, is it somehow converted or use it that I'm not aware of?

mattinbits replied:
I am happy to revert to the original style, but I agree with @hntd187 the current cache implementation is not futures based and the `get(...)` in the original implementation does not appear atomic to me. There's no synchronisation and `java.util.LinkedHashMap` is not a collection designed for concurrency.

Since we are using Akka, I think the appropriate way to get an atomic futures based cache would be to encapsulate the JobCache in an Actor (for atomicity) and use the ask pattern (for futures). (not in this PR though :smile: )

mattinbits replied:
Actually, having looked again I realise why I switched to pattern matching. It is in order to handle the fact that the cache may yield either JobJarInfo or PythonJobInfo and pattern matching is the most readable way to handle the check.

mattinbits replied:
Have extended the type of the cache key to include `BinaryType`.

velvia replied:
Huh.  I guess we removed spray-caching at some point.  Let's add that
back.  Spray caching is exactly what you describe - Futures and Akka-based,
with atomic operations.

I'm also OK using JDK8's ConcurrentHashMap with the computeIfAbsent
methods, which are similar.  We'd have to be OK with being a JDK8-only
codebase....

On Fri, Aug 12, 2016 at 2:10 AM, mattinbits <notifications@github.com>
wrote:

> In job-server/src/spark.jobserver/JobCache.scala
> <https://github.com/spark-jobserver/spark-jobserver/pull/553#discussion_r74561596>
> :
>
> >     * @param classPath the fully qualified name of the class/object to load
> >     */
> >    def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
> > -    cache.get((appName, uploadTime, classPath), {
> > -      import akka.pattern.ask
> > -      import scala.concurrent.Await
> > -
> > -      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
> > -      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
> > -      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
> > -      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
> > -      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
> > -      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
> > -      JobJarInfo(constructor, classPath, jarFilePath)
> > -    })
> > +    cache.get((appName, uploadTime, classPath)) match {
>
> I am happy to revert to the original style, but I agree with @hntd187
> <https://github.com/hntd187> the current cache implementation is not
> futures based and the get(...) in the original implementation does not
> appear atomic to me. There's no synchronisation and
> java.util.LinkedHashMap is not a collection designed for concurrency.
>
> Since we are using Akka, I think the appropriate way to get an atomic
> futures based cache would be to encapsulate the JobCache in an Actor (for
> atomicity) and use the ask pattern (for futures). (not in this PR though
> 😄 )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/553/files/b4ea3b476442b188582c22f86fe5e18e504de3b4#r74561596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABA321EkQkhl1RkfWL-0BOQPRzdGyINzks5qfDhpgaJpZM4JZq4L>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

velvia commented on job-server/src/spark.jobserver/JobManagerActor.scala:
```diff
@@ -216,14 +216,16 @@ class JobManagerActor(contextConfig: Config) extends InstrumentedActor {
-    if (!lastUploadTime.isDefined) return failed(NoSuchApplication)
+    val lastUploadTimeAndType = resp.uploadTimeAndType
+    if (!lastUploadTimeAndType.isDefined) return failed(NoSuchApplication)
+    val(lastUploadTime, binaryType) = lastUploadTimeAndType.get
```
space between val and (

velvia commented on job-server/src/spark.jobserver/io/JobDAO.scala:
```diff
@@ -2,16 +2,50 @@ package spark.jobserver.io
+  def name: String
+  def mediaType: MediaType
+}
+object BinaryType {
```
Blank line before this line for separation

velvia commented on job-server/src/spark.jobserver/io/JobDAO.scala:
```diff
@@ -2,16 +2,50 @@ package spark.jobserver.io
+}
+object BinaryType {
+  case object Jar extends BinaryType {
+    val extension: String = "jar"
```
It's customary to leave out the `: String`, but no biggie

velvia commented:
@mattinbits looked through most of it.   Looks good for the most part.  The most critical comment is about the `JobCache`, I believe your change makes the cache updates nonatomic and not thread safe.  Other comments are mostly cosmetic.  

Really awesome job and look forward to merging it soon.

mattinbits commented:
@velvia Thanks for the thorough review. I've addressed the comments except for around `convertRawResult` since I think we need to decide what approach to take. I've reverted the JobCache stuff to using the `get(,)` method used originally. I think the discussion around alterations to that belongs as a separate issue.

noorul commented on ci/install-python-dependencies.sh:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+pip install --user pyhocon
+pip3 install --user pyhocon
```
Let us ensure that we are following pep8

pip install pep8

noorul commented on .travis.yml:
```diff
@@ -1,14 +1,22 @@
+  - ci/install-spark.sh
 script:
-  - sbt clean coverage test coverageReport
+  - sbt clean coverage testPython test coverageReport
```
- find . -name \*.py -exec pep8 {} +

mattinbits commented:
Conforming to Pep8 and Pep8 check in Travis done.

noorul commented:
I will leave it to Evan to merge. LGTM

velvia commented:
@mattinbits just commented about the convertRawResults.  If that is addressed (with native JsonUtils changes) I think we should be good to go.

velvia commented:
@mattinbits LGTM.   Anything else you want to push before we merge it?

mattinbits commented:
@velvia No this is good to go from my PoV. Do you want me to squash it to a single commit?

velvia commented:
@mattinbits thanks, merging now.
